### PR TITLE
oslo-config: fix build

### PR DESCRIPTION
**Pull request by makefu**
Also updated the following dependencies:
keystoneauth1: 3.1.0 -> 3.2.0
  needed to disable test due to circular dependencies towards oslo-config
oslo-i18n: 2.7.0 -> 3.18.0
  update required to not have oslo-config as a build dependency
oslotest: 1.12.0 -> 2.18.0
os-client-config: 1.8.1 -> 1.28.0
  needed to disable testing due to circular dependency with oslotest
mox3: 0.11.0 -> 0.23.0
  for py36: needed to disable testing due to LOCALE flag ValueError
debtcollector: 0.9.0 -> 1.17.0

*update*:
oslo-service: 0.10.0 -> 1.26.0
  needs to disable tests due to network errors when importing eventlet
  for tests ( socket.getprotobyname('tcp') -> no such protocol )
eventlet: 0.17.4 -> 0.20.0
  cannot update to 0.21.0 due to version pinning ( < 0.21.0 ) of oslo-service
monotonic: 0.4 -> 1.3
oslo-serialization: 1.10.0 -> 2.20.0
oslo-utils: 2.6.0 -> 3.29.0
oslo-concurrency: 2.7.0 -> 3.22.0
oslo-log: 1.12.1 -> 3.31.0
oslo-context: 0.7.0 -> 2.18.1
routes: 1.12.3 -> 2.4.1
webob: 1.4.1 -> 1.7.3

extra packages:
requestsexceptions: init at 1.3.0
  removed `hacking` reference in `test-dependencies.txt` as it is not a test dependency

###### Motivation for this change

oslo-config broke (and never got fixed) with the update introduced in d58e1f1c7b3e955d3ee1194ea96c574338b47b97 . It got marked as broken 67651d80bc8baaf09ab91fec8ea423e09107ed8f and is now in the process of being deleted via the PR to remove *all things openstack* #29491 .
oslo-config however is a dependency for a number of different other packages. My main reason to have working tests it in nixpkgs is for sqlalchemy_migrate ( -> tempest-lib -> oslo-config ).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -18722,6 +18725,31 @@ in {
...
     };
   };
+  requestsexceptions = buildPythonPackage rec {
+    pname = "requestsexceptions";
+    version = "1.3.0";
+    name = "${pname}-${version}";
+
+    propagatedBuildInputs = with self; [ pbr ];
+
+    # upstream hacking package is not required for testing
```
[the remark this answers is not in the document]
Comment (makefu) [reply pkgs/top-level/python-packages.nix]: The package does not provide any tests (yet). However the test will evaluate the python files for syntax and import errors

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -18722,6 +18725,31 @@ in {
       maintainer = maintainers.fridh;
     };
   };
+  requestsexceptions = buildPythonPackage rec {
```
[the remark this answers is not in the document]
Comment (makefu) [reply pkgs/top-level/python-packages.nix]: an specific reason not to keep it in python-packages ? it is in fact only a build dependency for `os-client-config`.

[the remark this answers is not in the document]
Comment (Mic92) [reply pkgs/top-level/python-packages.nix]: We want to move all packages out of this file to make maintenance easier. But it is a lot of labour work so it will take some time.

[the remark this answers is not in the document]
Comment (makefu) [reply pkgs/top-level/python-packages.nix]: okay, i will rewrite the commit and put requestsexceptions into a separate file

File: pkgs/top-level/python-packages.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (makefu) [reply pkgs/top-level/python-packages.nix]: @nlewo as described in https://pypi.python.org/pypi/hacking/ the project is only for checking the openstack styleguide. These tests however are not of functional nature and therefore can be omitted. I think we should not start hot-fixing style-issues in packages for the sake of pleasing the pedantic style checker.

[the remark this answers is not in the document]
Comment (nlewo) [reply pkgs/top-level/python-packages.nix]: :) 
It's fine. It was just an observation.

**Comment by makefu** [comment]
@nlewo @Mic92 ok i've rebased and now the new package resides in python-modules.

**Comment by FRidh** [comment]
Good that you are having a look at this. I strongly recommend you add yourself as maintainer to packages that are important to you, be it directly or indirectly.

**Comment by FRidh** [comment]
`python.pkgs3.oslo-service` fails. Packages that can now be build again, but fail, should either be fixed or marked as broken.

**Comment by makefu** [comment]
@FRidh is there a command which builds all the packages dependent on the packages i touched? i tried `nox-review pr` but it always crashes with 
```
subprocess.CalledProcessError: Command '['git', 'checkout', '--quiet', '35f205a4b624bceca7c53b9c19ddac5f37a5ae4c']' returned non-zero exit status 1
```

I will add myself to the list of maintainers for the tools i need in nixpkgs 👍

**Comment by makefu** [comment]
ok, I've finished my updating frenzy (and soldered some cables while waiting for the build).

Is there a good way to test if all dependent pachages still build?

**Comment by FRidh** [comment]
@makefu sometimes you need to clear `~/.nox` when that happens. I use my own script, [niff](https://github.com/FRidh/niff), to determine what packages were changed.

You could run e.g. `niff pr 29602` in which case it compares your branch to the branch the PR is made against. Therefore, don't forget to rebase.

To also build the changes
```
nix-shell -p niff # its available in unstable and 17.09
nix-build -k $(niff pr 29602 --output=attrs)
```

**Comment by nlewo** [comment]
@Mic92 would it be possible to backport the oslo fix (commit 7251699) to 17.09 ?

**Comment by Mic92** [comment]
@makefu if I backport oslo, what else I need to backport?

**Comment by FRidh** [comment]
17.09 currently: https://hydra.nixos.org/eval/1396332
17.09 with this merge commit: https://hydra.nixos.org/eval/1396335

**Comment by Mic92** [comment]
So this is pyramid and something else?

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -7333,6 +7350,9 @@ in {
       webtest
       zope_component
       zope_interface
+      plaster
+      plaster-pastedeploy
+      hupper
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: why are these added to `buildInputs`? Seems to be needed at runtime (install_requires).

**Comment by copumpkin** [comment]
See for example https://hydra.nixos.org/job/nixpkgs/trunk/python27Packages.hupper.x86_64-darwin/all?page=1, although it's unclear why it's failing. It just doesn't seem to notice changes... 😦

**Comment by copumpkin** [comment]
I fixed `hupper` in https://github.com/NixOS/nixpkgs/commit/73c30b68e9b5f8b95ce93d5a593780c4f477f9de so I think we're all set now.

**Comment by makefu** [comment]
@copumpkin thanks for the investigation and fix, i really appreciate it as i do not have any darwin builder to test 👍

**Comment by copumpkin** [comment]
No problem, thanks for the initial update 😄
